### DISCUSSION TITLE
How to pick a line?

### DISCUSSION THREAD

**Discussion by carolrz** [pyvista/pyvista#2565] [Q&A] [answered]
Hi, I want to implement lines-picking with a mesh only built from lines. But `enable_cell_picking` doesn't work. 
For example, two lines are created as followings. And the exception _**"KeyError: 'Data array (original_cell_ids) not present in this dataset."**_ is raised while pressing P.  Did I misunderstand something? I am using pyvista 0.34.0 and python 3.8.3.
```
import numpy as np
import pyvista as pv

v = np.array([[0,0,0], [1,1,1], [2,2,2], [3,3,3], [0,1,2], [0,2,4]]).astype('float')
lines =[4, 0, 1, 2, 3, 3, 0, 4, 5]

pl = pv.Plotter()
mesh = pv.PolyData(var_inp=v, lines=lines, n_lines=2)
pl.add_mesh(mesh, pickable=True)
pl.enable_cell_picking(through=False, color='r')
pl.show()
```

BTW, I have noticed this [issue ](https://github.com/pyvista/pyvista-support/issues/127), which example is running with error **"AttributeError: 'PolyData' object has no attribute 'wireframe'"**. And it is also useless for picking an edge to set `stype='wireframe'` in `add_mesh`, which only supports picking a triangle surface.

**Comment by tkoyama010** [comment, marked as answer]
I could reproduce the error in my environment.
```python
pl.enable_cell_picking(through=True, color='r')
```
When setting `through` in `enable_cell_picking` True, it works correctly.

**Reply by carolrz** [reply]
Thx for your reply. With setting `through=True`, the error is not shown but the color is still not changed. It looks like a bug in picking cell visualization.

https://user-images.githubusercontent.com/24415759/167074247-27bc6a83-41a5-481e-b374-8c60bcb7f600.mp4

**Reply by tkoyama010** [reply]
The picking box is not big enough. Please specify the box so that the entire subject fits in the box.

**Reply by tkoyama010** [reply]
Note:
Visible cell picking (through=False) will only work if the mesh is displayed with a 'surface' representation style (the default).
https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.enable_cell_picking.html

**Reply by carolrz** [reply]
Thanks a lot! Bigger enough box works perfectly. 👍  Besides, I have another question. There are some long lines neighboring each other in my mesh. So it's not easy to drag a precise box to pick one. Is it possible to pick the whole line by a part of it/some points or any other method?

**Reply by tkoyama010** [reply]
Currently, the picked range node is used to create an actor after pickup. Therefore, it is not possible to select and pick up a part of the range.
https://github.com/pyvista/pyvista/blob/6bb517cd96f70b294e68efa0cfff441c222ee04c/pyvista/plotting/picking.py#L370-L408
We can [suggest adding a feature](https://github.com/pyvista/pyvista/issues/new?assignees=&labels=feature-request&template=feature-request.md&title=) if you want.

**Reply by carolrz** [reply]
I see. And it's OK for me. Thanks again for your help :)

**Reply by carolrz** [reply]
Hi, tkoyama, there is another problem. I also want to show the original vertex idx when picking the line. But this callback function gives different results under repeatedly picking, also the wrong idx, which should be (0, 4, 5). The picked path is an UnstructuredGrid object and I can't find the corresponding method but `cell_connectivity`, which may not be used correctly.  Example code is as follows. Would you mind giving me some tips?

```
import numpy as np
import pyvista as pv

def show_idx(path):
    pl.remove_actor(actor='path_actor')
    if path is None:
        return
    path_idx = path.cell_connectivity
    pl.add_text(f'path idx: {path_idx}', position='left_edge', color='blue', font_size=10, name='path_actor')
    
v = np.array([[0,0,0], [1,1,1], [2,2,2], [3,3,3], [0,1,2], [0,2,4]]).astype('float')
lines =[4, 0, 1, 2, 3, 3, 0, 4, 5]

pl = pv.Plotter()
mesh = pv.PolyData(var_inp=v, lines=lines, n_lines=2)
pl.add_mesh(mesh, pickable=True)
pl.enable_cell_picking(mesh=mesh, callback=show_idx, through=True, color='r')
pl.show()
```

![image](https://user-images.githubusercontent.com/24415759/167339630-0c10a8e2-7a2b-4c71-bb1c-0425b3de3944.png)
